mapproxy/source/mapnik.py

```diff
@@ -93,24 +96,42 @@ def render(self, query):
         else:
             return self.render_mapfile(mapfile, query)
 
-    def map_obj(self, mapfile):
-        if not self._cache_map_obj:
-            m = mapnik.Map(0, 0)
-            mapnik.load_map(m, str(mapfile))
-            return m
+    def _create_map_obj(self, mapfile):
+        m = mapnik.Map(0, 0)
+        mapnik.load_map(m, str(mapfile))
+        return m
 
+    def map_obj(self, mapfile):
         # cache loaded map objects
-        # only works when a single proc/thread accesses this object
+        # only works when a single proc/thread accesses the map
         # (forking the render process doesn't work because of open database
         #  file handles that gets passed to the child)
-        if mapfile not in self._map_objs:
-            with self._map_objs_lock:
-                if mapfile not in self._map_objs:
-                    m = mapnik.Map(0, 0)
-                    mapnik.load_map(m, str(mapfile))
-                    self._map_objs[mapfile] = m
-
-        return self._map_objs[mapfile]
+        # segment the cache by process and thread to avoid interference
+        if self._cache_map_obj:
+            cachekey = None # renderd guarantees that there are no concurrency issues
```

Comment:
All MapnikSources will re-use the same Mapnik object (with the first loaded mapfile).

Reply:
That would have been a bug — thank you for catching it! The mapfile should still be part of the key.

Reply:
The mapfile is now part of the key.